File: FizzBuzz/src/fizzbuzz.ts
export class FizzBuzzError extends Error {
    constructor(message: string) {
        super(message)
    }
}

export class FizzBuzz {
    private DIVISORS = {
        Fizz: 3, Buzz: 5
    }

    readonly PRIMES_TO_100 = [
        2, 3, 5, 7, 11,
        13, 17, 19, 23, 29,
        31, 37, 41, 43, 47,
        53, 59, 61, 67, 71,
        73, 79, 83, 89, 97
    ];

    range(max: number): string {
        this.handleRangeErrors(max);

        return this.arrayOnRange(max)
                    .map(val => this.of(val))
                    .join(' ')
    }

    withDivisors(fizzDivisor: number, buzzDivisor: number): FizzBuzz {
        this.handleDivisorsErrors(fizzDivisor, buzzDivisor);

        this.DIVISORS.Fizz = fizzDivisor;
        this.DIVISORS.Buzz = buzzDivisor;
        return this;
    }

    private of(value: number): string {
        if (this.divisibleBy15(value)) {
            return 'FizzBuzz'
        }
        else if (this.divisibleBy3(value)) {
            return 'Fizz';
        }
        else if (this.divisibleBy5(value)) {
            return 'Buzz';
        }
        
        return value.toString();
    }

    private divisibleBy3(value: number): boolean {
        return value % 3 === 0;
    }

    private divisibleBy5(value: number): boolean {
        return value % 5 === 0;
    }

    private divisibleBy15(value: number): boolean {
        return this.divisibleBy3(value) && this.divisibleBy5(value);
    }

    private arrayOnRange(range: number): number[] {
        return [...Array(range).keys()].map(v => v+1);
    }

    private handleRangeErrors(max: number): void {
        if (max < 0) {
            throw new FizzBuzzError('Cannot fizzbuzz negative numbers');
        }
    }

    private handleDivisorsErrors(first: number, second: number): void {
        if (first === second) {
            throw new FizzBuzzError('Divisors must be distinct');
        }
        if (first === 1 || second === 1) {
            throw new FizzBuzzError('Divisors cannot be 1');
        }
        if (first < 0 || second < 0) {
            throw new FizzBuzzError('Divisors cannot be negative');
        }
        if (!this.isPrimeUpTo100(first) || !this.isPrimeUpTo100(second)) {
            throw new FizzBuzzError('Divisors must be primes');
        }
    }

    private isPrimeUpTo100(value: number): boolean {
        return this.PRIMES_TO_100.indexOf(value) < 0 
    }
}

File: FizzBuzz/src/main/java/FizzBuzz.java
import java.util.Arrays;
import java.util.OptionalInt;

class FizzBuzzError extends Exception {
    FizzBuzzError(String message){
        super(message);
    }
}

public class FizzBuzz {
    private int max;
    private int fizzDivisor;
    private int buzzDivisor;

    private int[] primeNumbers = {
        2, 3, 5, 7, 11,
        13, 17, 19, 23, 29,
        31, 37, 41, 43, 47,
        53, 59, 61, 67, 71,
        73, 79, 83, 89, 97
    };

    FizzBuzz(int max) throws FizzBuzzError {
        // By default fizzDivisor = 3 && buzzDivisor = 5
        this.handleRangeError(max);
        this.max = max;
        this.fizzDivisor = 3;
        this.buzzDivisor = 5;
    }

    FizzBuzz(int fizz, int buzz, int max) throws FizzBuzzError {
        this.handleRangeError(max);
        handleDivisorErrors(fizz, buzz);
        this.max = max;
        this.fizzDivisor = fizz;
        this.buzzDivisor = buzz;
    }

    public int getMax() {
        return max;
    }

    private void handleRangeError(int max) throws FizzBuzzError {
        if(max < 0)
            throw new FizzBuzzError("Cannot fizzbuzz negative numbers");
    }

    private void handleDivisorErrors(int fizz, int buzz) throws FizzBuzzError {
        if(fizz == 1 || buzz == 1)
            throw new FizzBuzzError("Divisor cannot be 1");

        if(fizz < 0 || buzz < 0)
            throw new FizzBuzzError("Divisor cannot be negative");

        if(fizz == buzz)
            throw new FizzBuzzError("Divisors must be distinct");

        if(!isPrimeNumber(fizz) || !isPrimeNumber(buzz))
            throw new FizzBuzzError("Divisor must be a prime");
    }

    private boolean isPrimeNumber(int number){
        OptionalInt foundPrime = Arrays.stream(primeNumbers).filter(prime -> prime == number).findFirst();
        return foundPrime.isPresent();
    }

    private StringBuilder buildFizzBuzz(){
        StringBuilder fizzBuzz = new StringBuilder();
        char separator = ' ';

        for(int index = 1; index <= max; index++){
            boolean printSeparator = index != max;

            if(isMultiple(index, fizzDivisor))
                fizzBuzz.append("Fizz");

            if(isMultiple(index, buzzDivisor))
                fizzBuzz.append("Buzz");

            if(!isMultiple(index, fizzDivisor) && !isMultiple(index, buzzDivisor))
                fizzBuzz.append(index);

            if(printSeparator)
                fizzBuzz.append(separator);
        }
        return fizzBuzz;
    }

    private boolean isMultiple(int number, int divisor){
        return number % divisor == 0;
    }

    @Override
    public String toString() {
        return buildFizzBuzz().toString();
    }
}
